openmls/src/group/core_group/new_from_welcome.rs

```diff
@@ -238,21 +213,100 @@ impl CoreGroup {
         let resumption_psk = group_epoch_secrets.resumption_psk();
         resumption_psk_store.add(public_group.group_context().epoch(), resumption_psk.clone());
 
-        let group = CoreGroup {
+        let group = StagedCoreWelcome {
             public_group,
             group_epoch_secrets,
             own_leaf_index,
             use_ratchet_tree_extension: enable_ratchet_tree_extension,
             message_secrets_store,
             resumption_psk_store,
+            verifiable_group_info,
+            group_secrets,
+            leaf_keypair,
         };
+
+        Ok(group)
+    }
+
+    /// Returns the [`LeafNodeIndex`] of the group member that authored the [`Welcome`] message.
+    pub fn welcome_sender_index(&self) -> LeafNodeIndex {
+        self.verifiable_group_info.signer()
+    }
+
+    /// Returns the [`LeafNode`] of the group member that authored the [`Welcome`] message.
+    pub fn welcome_sender(&self) -> Result<&LeafNode, LibraryError> {
+        let sender_index = self.welcome_sender_index();
+        self.public_group
+            .leaf(sender_index)
+            .ok_or(LibraryError::custom(
+                "no leaf with given welcome sender index exists",
+            ))
+    }
+
+    /// Consumes the [`StagedCoreWelcome`] and returns the respective [`CoreGroup`].
+    pub fn into_core_group<KeyStore: OpenMlsKeyStore>(
```

Comment:
I suppose you only want to derive the path secrets and generate the keypairs in this step because you assume the application has done some inspection before and might abort the process, in which case we would have done something expensive for no reason.

I see how that makes sense from an efficiency perspective, but it means that `StagedWelcome::into_group()` can now fail because the Welcome message was invalid. This is weird from an application's point of view, because usually all possible checks are done before the application does the inspection.

I don't have a super strong opinion about this, but I think the deviation from the current norm should be strongly motivated (e.g. explain how important the efficiency benefit is) and properly documented. Otherwise I'd say we align this with the rest of the processing and do this step in `new_from_welcome()`.

Reply:
Good point. I was able to move the operations that might fail based for malformed messages into `new_from_welcome`. We still have to return a `Result` because storing the key might fail, though.